src/MongoDB.Driver.Core/WriteConcern.cs

```diff
@@ -135,6 +135,13 @@ public static WriteConcern FromBsonDocument(BsonDocument document)
                 writeConcern = writeConcern.With(journal: j.ToBoolean());
             }
 
+            // seems hacky - addresses the specific new test added
```

Comment:
After reading the drivers ticket a bit more closely, I don't think this isn't hacky at all: this is the exact case that we need to guard against. @rstam, what do you think?

Reply:
It's unclear to me where this check should be made.

* Connection string parsing?
* WriteConcern constructor?
* this method?

The test of w and j itself seems valid. Just not sure this is the right place for it.

Reply:
The Python implementation of this ticket checks w and j in the WriteConcern constructor, and after looking through I think that's probably the best way to validate it, since that should cover the changes in both the connection string and document tests.

Reply:
My original reasoning for including the check here is that this is called in both the connection string and document test runner. Semantically, I agree it probably doesn't make sense to make the check here - however, I'm not sure ConnectionString Parse() is the best place to call it, since that isn't specific to write concern. When I tried putting the validation in the WriteConcern constructor, it didn't work since our implementation seems to construct default WriteConcern objects and then assigns w and journal values in FromBsonDocument.

Reply:
It's possible we could have two validation checks - one in ConnectionString Parse() (specifically in the ParseOptions helper) and one in FromBsonDocument. This would avoid the need to refactor the AssertValid and AssertInvalid functions from ConnectionStringTestRunner. The reason I originally chose FromBsonDocument as the location for the check is because both test runners call this function; however, in ConnectionStringTestRunner it's only called after Parse() is called (leading to confusion with the Asserts), while in DocumentTestRunner, FromBsonDocument is the main parse-like function.

Reply:
Having the test in `ConnectionString.Parse()` seems reasonable to me. Addressing @rstam 's concern about whether this is the correct place for the test of `w` and `j`, upon closer reflection, having the test execute in the constructor seems reasonable to me, leaving the `FromBsonDocument` logic as purely concerned with parsing the `BsonDocument` and constructing a `WriteConcern`. Given that `FromBsonDocument` calls `WriteConcern.With()` which calls the constructor, I think the exception should still be thrown as needed in the `DocumentTestRunner`--what do you think @ksadoff ?

Reply:
@vincentkam I moved the check to the constructor and it seems to pass the tests, so hopefully that's the best place for it. I think I was looking at a different implementation of the constructor before.

Reply:
SGTM! 😄